dot/sync/worker.go

```diff
@@ -1,95 +1,106 @@
-// Copyright 2021 ChainSafe Systems (ON)
+// Copyright 2023 ChainSafe Systems (ON)
 // SPDX-License-Identifier: LGPL-3.0-only
 
 package sync
 
 import (
-	"context"
-	"sync"
+	"errors"
+	"fmt"
 	"time"
 
-	"github.com/libp2p/go-libp2p/core/peer"
-
 	"github.com/ChainSafe/gossamer/dot/network"
-	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/libp2p/go-libp2p/core/peer"
 )
 
-// workerState helps track the current worker set and set the upcoming worker ID
-type workerState struct {
-	ctx    context.Context
-	cancel context.CancelFunc
+var ErrStopTimeout = errors.New("stop timeout")
 
-	sync.Mutex
-	nextWorker uint64
-	workers    map[uint64]*worker
+type worker struct {
+	status      byte
+	peerID      peer.ID
+	sharedGuard chan struct{}
+
+	punishment chan time.Duration
+	stopCh     chan struct{}
```

Comment:
not sure we need this `stopCh`. You can signal to stop the worker by closing `queue`.